### PR TITLE
Fixes to peep happiness, hunger, thirst, thoughts.

### LINKED ISSUE #3353: Unhappy guests don't get entertained
**OS:** Windows 10
**Version:** 0.0.5-develop
**Commit:** 423844c

Related to one of the points in #2574
Thread on openrct2.org: https://openrct2.org/forums/topic/1162-t/

Unhappy guests don't get any happiness out of rides, their happiness even drops while the vehicle is not moving, making them leave more angry when they leave the ride. Also exiting the ride does not give them any excitement, while for other guests it does.

- [ ] Reproducible in RCT2 (vanilla)?
- [ ] Multiplayer?

**Steps to reproduce:**
1. In the attached saved game, find any unhappy guests that want to go on The White Whale (click any of the cars of a train that's about to leave the station multiple times to find one quickly)
2. Check their happiness as the train moves

**Save game:**
[Unhappy guests.sv6.txt](https://github.com/OpenRCT2/OpenRCT2/files/226854/Unhappy.guests.sv6.txt)

### PR DESCRIPTION
**Pull request by zaxcav**
Rebased on #4655 which corrects peep happiness - all commits to this end in this PR reverted.
Fixes potential under/over-flow in peep_generate()
Fixes condition for running out of money thought.
Adds comments to the related areas of the peep code.

### CODE REVIEW

**Comment by janisozaur** [comment]
@zaxcav such changes will cause increased desyncs, please bump network version. Same goes for your other PR and it may be worth merging them at once to limit amount of times we need to change it.

**Comment by marijnvdwerf** [comment]
Thanks for taking the effort to add comments!

**Comment by zaxcav** [comment]
@janisozaur Yes, I was made aware of that with the previous PR with path finding fixes.
I don't know what you main devs have as a policy on how often you want the network version bumped.
I should have asked though.

I'll bump the network version in this PR.

**Comment by zaxcav** [comment]
Ha.  Typo in that last commit description.  "Dump" should be "Bump".  How do you change that?

**Comment by janisozaur** [comment]
```
git commit --amend
# fix your commitmsg
git push -f
```
We don't have any strict policy, but we try to keep desyncs to minimum, or else users complain. We are already lenient enough to let them continue playing after a desync has occurred, most games outright eject such clients.

**Comment by zaxcav** [comment]
@janisozaur Thanks.

**Comment by zaxcav** [comment]
I'm currently doing an empirical comparison of the effect on the happiness of peeps riding a coaster between RCT2 and OpenRCT2 to see if OpenRCT2 is now working the same as RCT2.  On the RCT2 side of things this means analysing a sequence of screenshots with the peep status open... time consuming.

**Comment by duncanspumpkin** [comment]
@zaxcav i'll check it against the original code it will be a bit quicker than that.

File: src/peep/peep.c [hunk 1]
```diff
@@ -1089,11 +1105,17 @@ static void sub_68F41A(rct_peep *peep, int index)
 						possible_thoughts[num_thoughts++] = PEEP_THOUGHT_TYPE_BATHROOM;
 					}
 
-					// Not sure why the happiness check is like that seems wrong to me
 					if (!(gParkFlags & PARK_FLAGS_NO_MONEY) &&
 						peep->cash_in_pocket <= MONEY(9, 00) &&
 						peep->happiness >= 105 &&
-						peep->happiness >= 70){
+						peep->energy >= 70){
```
Comment (duncanspumpkin) [inline src/peep/peep.c]: it is peep->happiness for both checks in the original. Even though it is odd.

Comment (IntelOrca) [inline src/peep/peep.c]: original bug maybe?

Comment (duncanspumpkin) [inline src/peep/peep.c]: I wouldn't want to fix it right now though until we work out why peeps seem to be getting angry quicker.

Comment (zaxcav) [inline src/peep/peep.c]: So, original bug? Or toss out this commit?

Comment (zaxcav) [inline src/peep/peep.c]: That particular change is related to the running out of money thought - commit 43ef6d9.
The comparison I'm doing is related to commit cbc8db0.  If you have some way to check this easily, that would be fantastic!

**Comment by zaxcav** [comment]
So I've finished comparing with RCT2.  In OpenRCT2 the guests are now getting a higher ride satisfaction bonus than in RCT2.
I've got some theories about where the discrepancies could come from which I'll look into later.

**Comment by duncanspumpkin** [comment]
@zaxcav if you need me to check the original just point to where you want it compared and I can double check.

**Comment by zaxcav** [comment]
I finished empirically comparing the happiness of a peep going on a ride in RCT2 vs OpenRCT2.
Commit 7f58c89 more or less aligns the happiness between the two.
I did notice in the code that the happiness growth is sometimes higher for lower levels of ride satisfaction and lower for higher levels of ride satisfaction...
This sounds like a bug in the original to me.  Commit 573f899 simply reallocates the original bonus amounts according to ride satisfaction levels, so that happiness will grow more with higher ride satisfaction levels than with lower satisfaction levels.
If there are strong objections to this we can reverse this commit.

I'll post some graphs of the comparisons shortly.

File: src/peep/peep.c [hunk 2]
```diff
@@ -956,6 +963,10 @@ static void sub_68F41A(rct_peep *peep, int index)
 	}
 
 	if ((index & 0x1FF) == (gCurrentTicks & 0x1FF)){
+		/* Affect of masking with 0x1FF here vs mask 0x7F,
```
Comment (Gymnasiast) [inline src/peep/peep.c]: Effect

**Comment by zaxcav** [comment]
Here are the graphs I promised.
In all comparisons the data is from 5 peeps of varying initial happiness riding the Great White Wail coaster from the savegame in #3353.
The RCT2 data is identical in each graph and is collected from a series of screenshots by counting the number of pixels in the happiness bar.  The screen resolution used meant that 1 pixel represented 2+ happiness units, hence the slight unevenness in the plot.

I collected OpenrRCT2 data (debug output of the raw values) with commit cbc8db0, then with commit 7f58c89 and lastly with commit 573f899.
I haven't collected data pre this PR, but could if there's sufficient interest.

There seems to be differing offsets in when the happiness growth starts.  I haven't looked into where this comes from... while the same savegame is used for both RCT2 and OpenRCT2 the guests are still entering/walking up the queue, so possibly some random behaviour there results in the peeps entering the ride in a different order?

The first chart shows RCT2 vs OpenRCT2 for commit cbc8db0. Here it is obvious that peep happiness grows far too much in comparison with RCT2.
happiness1

The second chart shows RCT2 vs OpenRCT2 with commit 7f58c89 (in addition to cbc8db0).  Happiness levels more or less align. Guest 14 does seem to peak ever so slightly lower though.
happiness2

The third chart shows RCT2 vs OpenRCT2 with commit 573f899 (in addition to 7f58c89 and cbc8db0). The reallocated happiness growth means that peeps with higher ride satisfaction (which includes those I collected data for) get slightly happier from the ride as a result (other peeps with lower satisfaction won't get quite as happy from the ride, but none of the peeps used demonstrate this case).
happiness3

File: src/peep/peep.c (hunk 2 above)
Comment (zaxcav) [inline src/peep/peep.c]: Fixed

File: src/peep/peep.c [hunk 3]
```diff
@@ -1059,6 +1070,11 @@ static void sub_68F41A(rct_peep *peep, int index)
 		}
 
 		if ((index & 0x3FF) == (gCurrentTicks & 0x3FF)){
+			/* Affect of masking with 0x3FF here vs mask 0x1FF,
```
Comment (zaxcav) [inline src/peep/peep.c]: Fixed

**Comment by duncanspumpkin** [comment]
I will take a look tonight to see if any of this is due to bugs during implementation.

**Comment by IntelOrca** [comment]
@duncanspumpkin 

> I will take a look tonight to see if any of this is due to bugs during implementation.

Did you manage to do this, shall this be merged anyway?

**Comment by janisozaur** [comment]
@duncanspumpkin let me know if you OK this change, I will merge it together with the other two @zaxcav's PRs (#4576, #4529), so we only change network version once on these. If you have any comments for these PRs, please let me know.

**Comment by duncanspumpkin** [comment]
I've still not had time to look at this. Will try again tonight.

File: src/peep/peep.c [hunk 4]
```diff
@@ -6980,23 +7019,39 @@ rct_peep *peep_generate(int x, int y, int z)
...
+	 * to any value 0..255. */
+	peep->happiness = gGuestInitialHappiness;
+	/* Assume a default initial happiness of 0 is wrong and set
+	 * to 128 (50%) instead. */
 	if (gGuestInitialHappiness == 0)
-		happiness += 0x80;
-
-	peep->happiness = happiness;
-	peep->happiness_growth_rate = happiness;
+		peep->happiness = 0x80;
```
Comment (duncanspumpkin) [inline src/peep/peep.c]: change to 128

**Comment by duncanspumpkin** [comment]
Argh this is a complete mess since the refactoring. I'm trying to work out what has gone wrong but there looks to be a few bugs introduced when it was refactored. This is the original function

```
static void peep_on_enter_ride(rct_peep *peep, int rideIndex)
{
	rct_ride *ride;
	ride_rating minIntensity, maxIntensity;
	ride_rating minNausea, maxNausea;
	uint16 satisfactionFlags;

	ride = GET_RIDE(rideIndex);
	peep->flags &= ~PEEP_FLAGS_RIDE_SHOULD_BE_MARKED_AS_FAVOURITE;
	if (ride->excitement == (ride_rating)0xFFFF) {
		satisfactionFlags = 0x1FF;
	} else {
		satisfactionFlags = 0;
		maxIntensity = (peep->intensity >> 4) * 100;
		minIntensity = (peep->intensity & 0xF) * 100;
		if (maxIntensity <= ride->intensity || minIntensity >= ride->intensity) {
			satisfactionFlags |= (1 << 1);
		}
		maxIntensity -= peep->happiness * 2;
		minIntensity += peep->happiness;
		if (maxIntensity <= ride->intensity || minIntensity >= ride->intensity) {
			satisfactionFlags |= (1 << 4);
		}
		maxIntensity -= peep->happiness * 2;
		minIntensity += peep->happiness;
		if (maxIntensity <= ride->intensity || minIntensity >= ride->intensity) {
			satisfactionFlags |= (1 << 7);
		}

		minNausea = NauseaMinimumThresholds[(peep->nausea_tolerance & 3)];
		maxNausea = NauseaMaximumThresholds[(peep->nausea_tolerance & 3)];
		if (maxNausea <= ride->nausea || minNausea >= ride->nausea) {
			satisfactionFlags |= (1 << 2);
		}
		maxNausea -= peep->happiness * 2;
		minNausea += peep->happiness;
		if (maxNausea <= ride->nausea || minNausea >= ride->nausea) {
			satisfactionFlags |= (1 << 5);
		}
		maxNausea -= peep->happiness * 2;
		minNausea += peep->happiness;
		if (maxNausea <= ride->nausea || minNausea >= ride->nausea) {
			satisfactionFlags |= (1 << 8);
		}
		satisfactionFlags |= (1 << 9);
	}

	if (!(RCT2_GLOBAL(RCT2_ADDRESS_PARK_FLAGS, uint32) & PARK_FLAGS_NO_MONEY)) {
		if (ride->value != 0xFFFF) {
			satisfactionFlags &= ~(1 << 9);
			satisfactionFlags |= (1 << 10);
			if ((money16)ride->value < ride->price) {
				satisfactionFlags &= ~(1 << 10);
				satisfactionFlags |= (1 << 9);
				if ((ride->value + ((ride->value * peep->happiness) / 256)) < ride->price) {
					satisfactionFlags &= ~(1 << 9);
				}
			}
		}
	}

	uint8 satisfactionVar = 0;
	if (satisfactionFlags & (1 << 9)) satisfactionVar += 15;
	if (satisfactionFlags & (1 << 10)) satisfactionVar += 40;
	if ((satisfactionFlags & (1 << 9)) && (satisfactionFlags & (1 << 10))) satisfactionVar -= 45;
	if ((satisfactionFlags & 6) == 6) {
		satisfactionVar += 70;
	} else {
		if (satisfactionFlags & 6) satisfactionVar += 15;
		if ((satisfactionFlags & 0x30) == 0x30) {
			satisfactionVar += 35;
		} else {
			if (satisfactionFlags & 0x30) satisfactionVar += 10;
			if ((satisfactionFlags & 0x180) == 0x180) {
				satisfactionVar += 10;
			} else {
				satisfactionVar -= 60;
			}
		}
	}

	if (peep->time_in_queue <= 750) {
		satisfactionVar += 10;
	} else if (peep->time_in_queue >= 2250) {
		satisfactionVar -= 10;
		if (peep->time_in_queue >= 4500) {
			satisfactionVar -= 25;
		}
	}

	if (peep_has_ridden(peep, peep->current_ride)) satisfactionVar += 10;
	peep_set_has_ridden(peep, peep->current_ride);
	if (peep->no_of_rides < 255) peep->no_of_rides++;

	if (peep_has_ridden_ride_type(peep, ride->type)) satisfactionVar += 10;
	peep_set_has_ridden_ride_type(peep, ride->type);

	uint8 unkExcitementValue = clamp(0, (ride->excitement / 4) + peep->happiness, 255);
	if (unkExcitementValue >= peep->var_FA) {
		if (peep->happiness >= 160 && peep->happiness_growth_rate >= 160) {
			peep->var_FA = unkExcitementValue;
			peep->flags |= PEEP_FLAGS_RIDE_SHOULD_BE_MARKED_AS_FAVOURITE;
		}
	}

	uint8 satisfaction = 0;
	if (satisfactionVar >= 0) {
		satisfaction++;
		if (satisfactionVar >= 20) {
			satisfaction++;
			if (satisfactionVar >= 40) {
				satisfaction++;
			}
		}
	}
	ride_update_satisfaction(ride, satisfaction);
	peep->happiness_growth_rate = clamp(0, satisfactionVar + peep->happiness_growth_rate, 255);

	uint32 nauseaMultiplier = clamp(64, 256 - peep->happiness_growth_rate, 200);
	uint32 nauseaGrowthRateChange = (ride->nausea * nauseaMultiplier) / 512;
	nauseaGrowthRateChange *= max(128, peep->hunger);
	nauseaGrowthRateChange /= 128;
	nauseaGrowthRateChange *= 2;
	nauseaGrowthRateChange >>= (peep->nausea_tolerance & 3);
	peep->nausea_growth_rate = (uint8)clamp(0, peep->nausea_growth_rate + nauseaGrowthRateChange, 255);
}
```

I think we will need to re-refactor it from this as its too hard to work out where the bugs are because it was done all in one big commit.

**Comment by duncanspumpkin** [comment]
I suggest refactoring the value calculation part of the satisfaction to the following:

```
/* rct2: 0x00695555 */
sint16 calculate_ride_value_satisfaction(rct_peep* peep, rct_ride* ride) {
	if (gParkFlags & PARK_FLAGS_NO_MONEY)){
		return -30;
	}

	if (ride->value == 0xFFFF) {
		return -30;
	}

	money16 ridePrice = ride_get_price(ride);
	if (ride->value >= ridePrice) {
		return -5;
	}

	if ((ride->value + ((ride->value * peep->happiness) / 256)) >= ridePrice) {
		return -30;
	}

	return 0;
}
```
The rest of it will need similar refactoring.

**Comment by zaxcav** [comment]
At first glance I must admit that, while I can see the correlation in the guards to the original function, I don't follow the return values in the above.
If a refactor is needed here, I assume it's best to reject this PR.

**Comment by duncanspumpkin** [comment]
@zaxcav can you remove the changes to the satisfaction in this PR. I'll refactor it in a different PR then we can retest to see if fixes the issues. Keep this PR for your comments and your improvements to the init values.

**Comment by zaxcav** [comment]
Okay.  I'll rollback only the satisfaction changes as suggested. I should be able to do this today.

**Comment by duncanspumpkin** [comment]
Can you rebase this and it will be fine to merge

**Comment by zaxcav** [comment]
Rebased.  Should be good to go.

**Pull request merged**